[HTML5 2.0] Presentation Manager

oswaldoacauan opened the pull request:
In favor of #4371 since it was merged in the wrong branch.

OZhurbenko commented:
Shouldn't #4371 be reverted in this case?
@antobinary

capilkey commented:
#4371 should be reverted so there aren't any merge issues when we go to merge the 2.0 branch in.

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/component.jsx:
```diff
@@ -0,0 +1,458 @@
+  },
+  dismissDesc: {
+    id: 'app.presentationUploder.dismissDesc',
+    defaultMessage: 'Closes and discarts your changes',
```
Close the modal window and discard your changes

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/component.jsx:
```diff
@@ -0,0 +1,458 @@
+  },
+  browseFilesLabel: {
+    id: 'app.presentationUploder.browseFilesLabel',
+    defaultMessage: 'or browse for files',
```
Maybe just attach this part to the dropzoneLabel.defaultMessage?
This is just half of the message,

oswaldoacauan replied:
Each phrase has a different style. Didn't find a better way to do it without two messages, any ideas?

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/component.jsx:
```diff
@@ -0,0 +1,458 @@
+    conversion: PropTypes.object,
+    upload: PropTypes.object,
+  })).isRequired,
+};
```
```
intl: PropTypes.shape({
  formatMessage: PropTypes.func.isRequired,
}).isRequired,
```

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/container.jsx:
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { createContainer } from 'meteor/react-meteor-data';
+
+import Service from './service';
```
`PresentationUploaderService`

oswaldoacauan replied:
We only prefix the service name when importing outside its pod.

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/actions-bar/actions-dropdown/component.jsx:
```diff
@@ -9,6 +10,17 @@ import DropdownContent from '/imports/ui/components/dropdown/content/component';
 import DropdownList from '/imports/ui/components/dropdown/list/component';
 import DropdownListItem from '/imports/ui/components/dropdown/list/item/component';
 
+import PresentationUploderContainer from '/imports/ui/components/presentation/presentation-uploader/container';
```
typo - `PresentationUploaderContainer`

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/container.jsx:
```diff
@@ -0,0 +1,29 @@
+
+const PresentationUploaderContainer = props => (
+  <PresentationUploader {...props}>
+    {props.children}
```
I couldn't find the case where this component has `children`. Probably this can be removed.

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/service.js:
```diff
@@ -0,0 +1,121 @@
+import Auth from '/imports/ui/services/auth';
+import { makeCall } from '/imports/ui/services/api';
+
+const CONVERSION_TIMEOUT = 300000;
```
This should be in the config, in my opinion.

oswaldoacauan replied:
Will leave this change for the next iteration.

OZhurbenko replied:
Add this to the list of issues?

OZhurbenko commented on bigbluebutton-html5/imports/ui/components/presentation/presentation-uploader/service.js:
```diff
@@ -0,0 +1,121 @@
+const uploadPresentation = (file, meetingID, endpoint, onError, onProgress) => {
+  const data = new FormData();
+  data.append('presentation_name', file.filename);
+  data.append('Filename', file.filename);
```
We have such a mix of different naming conventions here
camelCase - `'fileUpload'`
underscore - `'presentation_name'` , `'is_downloadable'`
not sure what type this is - `'Filename'`

Probably this needs to be corrected.

OZhurbenko commented:
Here is the list of the bugs I was able to find at this moment:
1. - Select a file `name.pdf`
    - Set the checkbox to make it  `current`
    - Start.
It uploads and converts it, but doesn't make your newly uploaded presentation `current`.
2. Handle presentations with the same file names. If you try to upload a presentation with the same name currently - everything will break in html5.
3. It doesn't indicate in the presentation upload window which presentation is your current presentation.
4. You should be able to delete your current presentation, unless it's a default one.
5. If you go to upload presentation and try to set your actual current presentation again - everything hangs. You should disable "Start" button when your current presentation is Checked. Or display a disabled checked checkbox in front of your current presentation, so that users weren't able to check it and break the client.
6. If you steal a presenter from the user while he is in the presentation-upload window - nothing happens, the popup doesn't close.
7. Checkbox for file download is missing. The value should also be attached to the message.
8. We need to add a check [here](https://github.com/oswaldoacauan/bigbluebutton/blob/a3f74842b4684b2914f32dfe11196ef302830d6c/bigbluebutton-html5/imports/api/2.0/presentations/server/modifiers/changeCurrentPresentation.js#L45). If oldPresentation.id (_id?) == newPresentation.id (_id?) then return; to avoid the same [bug](https://github.com/bigbluebutton/bigbluebutton/pull/4431) as we had with Slide change.

OZhurbenko commented on bigbluebutton-html5/imports/api/2.0/presentations/server/methods/removePresentation.js:
```diff
@@ -0,0 +1,34 @@
+  check(requesterUserId, String);
+  check(presentationId, String);
+
+  const presenationToDelete = Presentations.findOne({
```
[comment](https://github.com/bigbluebutton/bigbluebutton/pull/4371#discussion_r139784297)

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/handlers/presentationCurrentChange.js:
```diff
@@ -0,0 +1,11 @@
+import { check } from 'meteor/check';
+import changeCurrentPresentation from '../modifiers/changeCurrentPresentation';
```
I don't agree with the `setCurrentPresentation`. We are not only setting the current to true, we are also setting the current one to false. So we are "changing" the current presentation to a new one, not just setting as current.

oswaldoacauan replied:
Since all events are `SetCurrent` I will rename so we follow akka.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/handlers/presentationCurrentChange.js:
```diff
@@ -0,0 +1,11 @@
+import changeCurrentPresentation from '../modifiers/changeCurrentPresentation';
+
+export default function presentationCurrentChange({ body }, meetingId) {
+  const { presentationId } = body;
```
Didn't find a example where we check the `body` itself since we have no control of its shape. Will add a check for object instead.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/methods/removePresentation.js:
```diff
@@ -0,0 +1,34 @@
+  const CHANNEL = REDIS_CONFIG.channels.toAkkaApps;
+  const EVENT_NAME = 'RemovePresentationPubMsg';
+
+  const { meetingId, requesterUserId } = credentials;
```
I'm just following the pattern of all other methods in the client, this is the exact current implementation in every method of the client.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/methods/sharePresentation.js:
```diff
@@ -0,0 +1,32 @@
+  const CHANNEL = REDIS_CONFIG.channels.toAkkaApps;
+  const EVENT_NAME = 'SetCurrentPresentationPubMsg';
+
+  const { meetingId, requesterUserId } = credentials;
```
I'm just following the pattern of all other methods in the client, this is the exact current implementation in every method of the client.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/eventHandlers.js:
```diff
@@ -2,8 +2,13 @@ import RedisPubSub from '/imports/startup/server/redis2x';
+RedisPubSub.on('PresentationPageGeneratedEvtMsg', handlePresentationConversionUpdate);
+RedisPubSub.on('PresentationConversionUpdateEvtMsg', handlePresentationConversionUpdate);
 RedisPubSub.on('PresentationConversionCompletedEvtMsg', handlePresentationChange);
+RedisPubSub.on('NewPresentationEvtMsg', handlePresentationChange);
```
https://github.com/bigbluebutton/bigbluebutton/blob/bbb-2x-mconf/bbb-common-message/src/main/scala/org/bigbluebutton/common2/msgs/PresentationMsgs.scala#L67

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/handlers/presentationChange.js:
```diff
@@ -1,45 +1,11 @@
-
 export default function handlePresentationChange({ header, body }) {
   const { meetingId } = header;
   const { presentation } = body;
```
Didn't find a example where we check the body itself since we have no control of its shape. Will add a check for object instead.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/handlers/presentationChange.js:
```diff
@@ -1,45 +1,11 @@
-  return Presentations.update(selector, modifier, cb);
-};
-
 export default function handlePresentationChange({ header, body }) {
```
We use the same method in the `NewPresentationEvtMsg`

oswaldoacauan replied:
Will rename it to `handlePresentationAdded` since the handler add the presentation.

oswaldoacauan commented on bigbluebutton-html5/imports/api/2.0/presentations/server/handlers/presentationConversionUpdate.js:
```diff
@@ -0,0 +1,84 @@
+// const CONVERSION_COMPLETED_KEY = 'CONVERSION_COMPLETED';
+
+export default function handlePresentationConversionUpdate({ body }, meetingId) {
+  const presentationId = body.presentationId;
```
Didn't find a example where we check the body itself since we have no control of its shape. Will add a check for object in

oswaldoacauan commented:
1. Fixed.
2. Know issue, will be fixed on next iterations.
3. The current UI does not indicate the "real" current presentation, good point to discuss.
4. Fixed.
5. Fixed, now the client will not hang.
6. Know issue, will be fixed on next iterations.
7. Know issue, theres no UI for that in the moment will be implemented in further iterations.
8. Fixed.

cc @OZhurbenko
